### ISSUE TITLE
hash function and proxies (this is complicated!)

### ISSUE THREAD

**Issue by jhprinz** [openpathsampling/openpathsampling#701]
This is more a discussion then really an issue. I realized that we use different concepts of equality and wanted an opinion. These are the different ways for equality we use and below I will explain why we need both and what we could do about it.

1. Compare features or attributes of two objects: This is intuitive and the pythonic way. If two things represent the same information they are equal. For example for volumes and ensembles we compare the string representation. if these are the same, we consider the objects equal. The problem is to find a suitable hash function.

2. Compare using the uuid, which should be really unique. So, if these are the same, we know objects are equal, but. If you make a deep copy, the UUID is different but the objects are equal in the first sense.

For most objects we use 2. but for some 1. So why use 2. if the other one is better?

Problem is that we also use shallow proxy objects, that contain no information but the UUID and a place to load it from. Of course, you would like to compare, a proxy with other objects and if the UUID is the same, they are equal. If you use 1. you will have to load every object for comparison. If you use 2. you do not. 

Also, in 2. you have a very good hash function which you need to place objects in a set or use as a key in a dict. If you put an object and its proxy in a set in 2. you have only one object. In 1. you have 2 objects. That is redundant for e.g. caching. Also in 1. the hash function might be very inefficient.

### Solutions

- Keep it the way it is and have special equality for some objects, Trajectory, Ensemble, Volume, etc
- Use only UUID hash. Equality will still work, but dicts will consider deep copies different, although eq might say different. 
- Change storage (at some point) and save the hash with the objects. This would be a big change, but technically pretty cool and advanced. I have not seen any of this considered in NoSQL DB (if that rings a bell)

For now I leave it the way it is, but especially for Trajectory and upcoming trajectory CVs this could be problematic...

**Comment by jhprinz** [comment]
Actually, I saw that we already use the UUID-hash for all objects except for trajectories. So we might switch to do that for trajectories as well. It is worth considering if that breaks caching for can_append, etc...

**Comment by dwhswenson** [comment]
> Actually, I saw that we already use the UUID-hash for all objects except for trajectories.

Not the case. See #700. We use string representation for volumes and ensembles.

With volumes and ensembles, I think this is used to identify equivalent ensembles from different simulation setups, so that can be useful.

I'm fine with adding UUID as the equality test for trajectories, though.

I'm not sure about our choice for samples, which is also different. We use a `collections.Counter` of the `SampleSet` object... which, I expect, uses the equality of trajectories and of ensembles within it. I guess that makes sense, but I think UUID is a better choice for `SampleSet` as well.

**Comment by dwhswenson** [comment]
Also, should we modify (or default) `__hash__` to be UUID? Maybe that isn't the case with everything, but there are some changes with `__hash__` and `__eq__` in Python 3. I had to update a few things to have an explicitly defined `__hash__`.

**Comment by jhprinz** [comment]
> Not the case. See #700. We use string representation for volumes and ensembles.

Yes, it is. I am only talking about the hash function! Not `__eq__`, there you are right. I just realized that for Ensembles we use the UUID __hash__ inherited from StorableObject. That means that things that need hash and eq will not work correctly for these. 

But that is another question. I think it might make sense to use the hash everywhere the same. But that will (as currently) lead to wrong behaviour for dicts and set, etc...

**Comment by dwhswenson** [comment]
> It is worth considering if that breaks caching for can_append, etc...

1. This relies on snapshot equality. Do proxies and full objects have the same UUID? (it would seem logical for them to)

2. Most of the time, it won't matter anyway -- you're not running `can_append` with proxies in your trajectory. And in any case, the caching in `can_append` is just a speed convenience; it will always end up with the right answer, although perhaps a bit slower.

> I think it might make sense to use the hash everywhere the same. But that will (as currently) lead to wrong behaviour for dicts and set, etc...

My intuition is to move toward `__hash__` as main underlying thing, and base `__eq__` on `__hash__`. Default of UUID makes sense; under most circumstances, making the UUID into the `__hash__` and making it the definition of equality is what we want. In special cases (where we want a different definition of equality, like with `Volume` and `Ensemble`), we would then need to override `__hash__`, which would have the effect of overriding `__eq__`.

I'm not sure on the expected behavior in Python 3, but that matches what I understand of it (since we no longer get automatic overrides of `__hash__` by overriding `__eq__`.)

Does that seem logical to you, @jhprinz?

**Comment by jhprinz** [comment]
> My intuition is to move toward __hash__ as main underlying thing, and base __eq__ on __hash__. Default of UUID makes sense; under most circumstances, making the UUID into the __hash__ and making it the definition of equality is what we want. In special cases (where we want a different definition of equality, like with Volume and Ensemble), we would then need to override __hash__, which would have the effect of overriding __eq__.

Yes, that works for me, only place where I am not sure it `Trajectories`.

Currently comparison __eq__ is inherited from list, so trajectories are equal when all the elements are equal. That is intuitive. The hash compares length, first and last element. That fits. But we get problems with proxies. Though I am not sure if we will have many trajectory proxies. Could be that we just do not have some, at least I did not find some.

Then this would not matter for now. Alright, then let's just consider updating hash functions for ensemble and volume, etc... We do not use proxies for these objects anyway.

**Comment by dwhswenson** [comment]
If `TrajectoryProxy` is a thing, then I think we should also do trajectory equality through UUID. (The proxy does have the same UUID as the full object, right?)

For the time being, I'm going to fix the issues in #700 with quick fix PR (about to start it). Those fixes might be overridden with a more complete approach (the tests should stay about the same).

**Comment by dwhswenson** [comment]
Closing here. See further discussion in #1080.
